File: BS.py
import os
import threading
import random
import time
from datetime import datetime

# Directory to store account files
ACCOUNTS_DIR = "accounts"

# Ensure the accounts directory exists
if not os.path.exists(ACCOUNTS_DIR):
    os.makedirs(ACCOUNTS_DIR)

# Dictionary to hold locks for each account
account_locks = {}
global_lock = threading.RLock()

# Lock for the central transaction log
transaction_log_lock = threading.Lock()

# Path for the central transaction log
CENTRAL_TRANSACTION_LOG = os.path.join(ACCOUNTS_DIR, "transactions.log")

def get_account_filepath(account_id):
    """Returns the file path for a given account."""
    return os.path.join(ACCOUNTS_DIR, f"{account_id}.txt")

def get_account_lock(account_id):
    """Retrieves or creates a lock for the given account."""
    with global_lock:
        if account_id not in account_locks:
            account_locks[account_id] = threading.RLock()
        return account_locks[account_id]

def read_balance(account_id):
    """Reads the balance of the specified account."""
    filepath = get_account_filepath(account_id)
    if not os.path.exists(filepath):
        print(f"Account {account_id} does not exist.")
        return None
    with open(filepath, "r") as file:
        try:
            balance = int(file.read().strip())
            return balance
        except ValueError:
            print(f"Invalid balance format for account {account_id}.")
            return None

def write_balance_atomic(account_id, new_balance):
    
    filepath = get_account_filepath(account_id)
    temp_filepath = filepath + ".tmp"
    try:
        with open(temp_filepath, "w") as temp_file:
            temp_file.write(f"{new_balance}\n")
        os.replace(temp_filepath, filepath)
        return True
    except Exception as e:
        print(f"Error writing balance for {account_id}: {e}")
        if os.path.exists(temp_filepath):
            os.remove(temp_filepath)
        return False

def log_transaction_atomic(operation_type, user_id, details, status):
    """Logs a transaction to the central transaction log."""
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    log_entry = f"{timestamp} | {operation_type} | {user_id} | {details} | {status}\n"
    with transaction_log_lock:
        try:
            with open(CENTRAL_TRANSACTION_LOG, "a") as log_file:
                log_file.write(log_entry)
        except Exception as e:
            print(f"Error logging transaction: {e}")

def create_account(account_id, initial_balance):
    """Creates a new account with the specified initial balance."""
    lock = get_account_lock(account_id)
    with lock:
        filepath = get_account_filepath(account_id)
        if os.path.exists(filepath):
            print(f"Account {account_id} already exists.")
            return
        success = write_balance_atomic(account_id, initial_balance)
        if success:
            log_transaction_atomic("Create Account", account_id, f"Initial balance: {initial_balance}", "Success")
            print(f"Account {account_id} created successfully.")
        else:
            log_transaction_atomic("Create Account", account_id, f"Initial balance: {initial_balance}", "Failed")
            print(f"Failed to create account {account_id}.")

def transfer(from_account, to_account, amount):
    """Transfers amount from from_account to to_account atomically."""
    if from_account == to_account:
        print("Cannot transfer to the same account.")
        log_transaction_atomic("Transfer", from_account, f"Attempted to transfer to self: {amount}", "Failed")
        return

    ordered_accounts = sorted([from_account, to_account])
    lock1 = get_account_lock(ordered_accounts[0])
    lock2 = get_account_lock(ordered_accounts[1])

    with lock1:
        with lock2:
            try:
                balance_from = read_balance(from_account)
                balance_to = read_balance(to_account)

                if balance_from is None or balance_to is None:
                    print(f"One or both accounts ({from_account}, {to_account}) do not exist.")
                    log_transaction_atomic("Transfer", from_account, f"To: {to_account}, Amount: {amount}", "Failed")
                    return

                if balance_from < amount:
                    print(f"Transfer failed: Insufficient funds in account {from_account}. Current balance: {balance_from}")
                    log_transaction_atomic("Transfer", from_account, f"To: {to_account}, Amount: {amount}", "Failed")
                    return

            
                new_balance_from = balance_from - amount
                new_balance_to = balance_to + amount

               
                success_from = write_balance_atomic(from_account, new_balance_from)
                success_to = write_balance_atomic(to_account, new_balance_to)

                if success_from and success_to:
                    log_transaction_atomic("Transfer", from_account, f"To: {to_account}, Amount: {amount}", "Success")
                    log_transaction_atomic("Transfer", to_account, f"From: {from_account}, Amount: {amount}", "Success")
                    print(f"Transferred {amount} from {from_account} to {to_account}.")
                else:
                    # If any write failed, rollback
                    if success_from:
                        # Attempt to rollback from_account
                        write_balance_atomic(from_account, balance_from)
                    if success_to:
                        # Attempt to rollback to_account
                        write_balance_atomic(to_account, balance_to)
                    log_transaction_atomic("Transfer", from_account, f"To: {to_account}, Amount: {amount}", "Failed")
                    log_transaction_atomic("Transfer", to_account, f"From: {from_account}, Amount: {amount}", "Failed")
                    print(f"Transfer of {amount} from {from_account} to {to_account} failed and has been rolled back.")

            except Exception as e:
                print(f"Error during transfer: {e}")
                log_transaction_atomic("Transfer", from_account, f"To: {to_account}, Amount: {amount}", "Failed")

def deposit(account_id, amount):
    """Deposits the specified amount into the account atomically."""
    lock = get_account_lock(account_id)
    with lock:
        try:
            balance = read_balance(account_id)
            if balance is None:
                print(f"Deposit failed: Account {account_id} does not exist.")
                log_transaction_atomic("Deposit", account_id, f"Amount: {amount}", "Failed")
                return
            new_balance = balance + amount
            success = write_balance_atomic(account_id, new_balance)
            if success:
                log_transaction_atomic("Deposit", account_id, f"Amount: {amount}", "Success")
                print(f"Deposited {amount} to {account_id}. New balance: {new_balance}")
            else:
                log_transaction_atomic("Deposit", account_id, f"Amount: {amount}", "Failed")
                print(f"Failed to deposit {amount} to {account_id}.")
        except Exception as e:
            print(f"Error during deposit: {e}")
            log_transaction_atomic("Deposit", account_id, f"Amount: {amount}", "Failed")

def withdraw(account_id, amount):
    """Withdraws the specified amount from the account atomically."""
    lock = get_account_lock(account_id)
    with lock:
        try:
            balance = read_balance(account_id)
            if balance is None:
                print(f"Withdrawal failed: Account {account_id} does not exist.")
                log_transaction_atomic("Withdraw", account_id, f"Amount: {amount}", "Failed")
                return
            if balance < amount:
                print(f"Withdrawal failed: Insufficient funds in account {account_id}. Current balance: {balance}")
                log_transaction_atomic("Withdraw", account_id, f"Amount: {amount}", "Failed")
                return
            new_balance = balance - amount
            success = write_balance_atomic(account_id, new_balance)
            if success:
                log_transaction_atomic("Withdraw", account_id, f"Amount: {amount}", "Success")
                print(f"Withdrew {amount} from {account_id}. New balance: {new_balance}")
            else:
                log_transaction_atomic("Withdraw", account_id, f"Amount: {amount}", "Failed")
                print(f"Failed to withdraw {amount} from {account_id}.")
        except Exception as e:
            print(f"Error during withdrawal: {e}")
            log_transaction_atomic("Withdraw", account_id, f"Amount: {amount}", "Failed")

def view_balance(account_id):
    """Displays the current balance of the account."""
    try:
        balance = read_balance(account_id)
        if balance is not None:
            print(f"Account {account_id} Balance: {balance}")
            log_transaction_atomic("View Balance", account_id, f"Balance: {balance}", "Success")
    except Exception as e:
        print(f"Error viewing balance: {e}")
        log_transaction_atomic("View Balance", account_id, f"Error: {e}", "Failed")

def generate_central_log():
    """Generates a central log of all account balances."""
    central_log_path = os.path.join(ACCOUNTS_DIR, "central_log.txt")
    with open(central_log_path, "w") as log_file:
        log_file.write("Central Log - Account Balances\n")
        log_file.write("-" * 50 + "\n")
        for account_file in os.listdir(ACCOUNTS_DIR):
            if account_file.endswith(".txt") and not account_file.endswith("_log.txt") and account_file != "transactions.log":
                account_id = account_file.replace(".txt", "")
                balance = read_balance(account_id)
                if balance is not None:
                    log_file.write(f"Account: {account_id}, Balance: {balance}\n")
    print(f"Central log generated at: {central_log_path}")

def user_operations(user_id, all_users):
    """Performs operations for a user: transfers, deposits, withdrawals, and view balance."""
    # Number of operations per user
    num_operations = 10

    for _ in range(num_operations):
        operation = random.choice(['transfer', 'deposit', 'withdraw', 'view_balance'])

        if operation == 'transfer':
            # Select a random target account excluding self
            target_account = random.choice([user for user in all_users if user != user_id])
            amount = random.randint(10, 1000)  # Random transfer amount
            transfer(user_id, target_account, amount)

        elif operation == 'deposit':
            amount = random.randint(10, 500)  # Random deposit amount
            deposit(user_id, amount)

        elif operation == 'withdraw':
            amount = random.randint(10, 500)  # Random withdrawal amount
            withdraw(user_id, amount)

        elif operation == 'view_balance':
            view_balance(user_id)

        # Random sleep to simulate real-world operations
        time.sleep(random.uniform(0.1, 0.5))

def main():
    # List of all user IDs
    user_ids = [f"User{i}" for i in range(1, 4)]  # User1 to User10

    # Initial balance for each account
    initial_balance = 1000

    # Pre-create all accounts before starting threads
    print("Creating all user accounts...")
    for user_id in user_ids:
        create_account(user_id, initial_balance)
    print("All accounts created.\n")

    # Start threads for each user
    threads = []
    for user_id in user_ids:
         thread = threading.Thread(target=user_operations, args=(user_id, user_ids))
         threads.append(thread)
         thread.start()

  

    # Wait for all threads to complete
    for thread in threads:
        thread.join()

    # Generate central log after all operations
    generate_central_log()
    print("All operations completed.")

if __name__ == "__main__":
    main()

File: BS.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <unistd.h>
#include <sys/stat.h>
#include <time.h>

// Directory to store account files
#define ACCOUNTS_DIR "accounts"

// Maximum number of accounts
#define MAX_ACCOUNTS 100

// Name of the transaction log file
#define TRANSACTION_LOG "transactions.log"

// Structure representing an account
typedef struct {
    char account_id[50];
    pthread_mutex_t lock;
} Account;

// Array of accounts
Account accounts[MAX_ACCOUNTS];
int account_count = 0;

// Mutex for managing the accounts array
pthread_mutex_t global_lock = PTHREAD_MUTEX_INITIALIZER;

// Mutex for the transaction log
pthread_mutex_t transaction_log_lock = PTHREAD_MUTEX_INITIALIZER;

// Path to the central transaction log
char central_transaction_log[100] = ACCOUNTS_DIR "/" TRANSACTION_LOG;

// Function to get the file path for an account
void get_account_filepath(const char *account_id, char *filepath) {
    sprintf(filepath, "%s/%s.txt", ACCOUNTS_DIR, account_id);
}

// Function to retrieve or create an account
Account* get_account(const char *account_id) {
    pthread_mutex_lock(&global_lock);
    for(int i = 0; i < account_count; i++) {
        if(strcmp(accounts[i].account_id, account_id) == 0) {
            pthread_mutex_unlock(&global_lock);
            return &accounts[i];
        }
    }
    // If account does not exist, create it
    if(account_count < MAX_ACCOUNTS) {
        strcpy(accounts[account_count].account_id, account_id);
        pthread_mutex_init(&accounts[account_count].lock, NULL);
        account_count++;
        pthread_mutex_unlock(&global_lock);
        return &accounts[account_count-1];
    }
    pthread_mutex_unlock(&global_lock);
    return NULL;
}

// Function to read the balance of an account
int read_balance(const char *account_id, int *balance) {
    char filepath[100];
    get_account_filepath(account_id, filepath);
    FILE *file = fopen(filepath, "r");
    if(file == NULL) {
        printf("Account %s does not exist.\n", account_id);
        return -1;
    }
    if(fscanf(file, "%d", balance) != 1) {
        printf("Invalid balance format for account %s.\n", account_id);
        fclose(file);
        return -1;
    }
    fclose(file);
    return 0;
}

// Function to write the balance to an account file atomically
int write_balance_atomic(const char *account_id, int new_balance) {
    char filepath[100];
    char temp_filepath[150];
    get_account_filepath(account_id, filepath);
    sprintf(temp_filepath, "%s.tmp", filepath);
    
    FILE *temp_file = fopen(temp_filepath, "w");
    if(temp_file == NULL) {
        printf("Error opening temporary file for account %s.\n", account_id);
        return -1;
    }
    fprintf(temp_file, "%d\n", new_balance);
    fclose(temp_file);
    
    // Replace the original file with the temporary file
    if(rename(temp_filepath, filepath) != 0) {
        printf("Error replacing file for account %s.\n", account_id);
        unlink(temp_filepath);
        return -1;
    }
    return 0;
}

// Function to log transactions atomically
void log_transaction_atomic(const char *operation_type, const char *user_id, const char *details, const char *status) {
    pthread_mutex_lock(&transaction_log_lock);
    FILE *log_file = fopen(central_transaction_log, "a");
    if(log_file == NULL) {
        printf("Error opening transaction log file.\n");
        pthread_mutex_unlock(&transaction_log_lock);
        return;
    }
    // Get current time
    time_t now = time(NULL);
    struct tm *t = localtime(&now);
    char timestamp[20];
    strftime(timestamp, sizeof(timestamp)-1, "%Y-%m-%d %H:%M:%S", t);
    
    fprintf(log_file, "%s | %s | %s | %s | %s\n", timestamp, operation_type, user_id, details, status);
    fclose(log_file);
    pthread_mutex_unlock(&transaction_log_lock);
}

// Function to create a new account
void create_account(const char *account_id, int initial_balance) {
    Account *account = get_account(account_id);
    if(account == NULL) {
        printf("Error creating account %s.\n", account_id);
        log_transaction_atomic("Create Account", account_id, "Initial balance", "Failed");
        return;
    }
    
    pthread_mutex_lock(&account->lock);
    char filepath[100];
    get_account_filepath(account_id, filepath);
    
    // Check if account file already exists
    FILE *file = fopen(filepath, "r");
    if(file != NULL) {
        printf("Account %s already exists.\n", account_id);
        fclose(file);
        pthread_mutex_unlock(&account->lock);
        log_transaction_atomic("Create Account", account_id, "Initial balance", "Failed");
        return;
    }
    
    // Write initial balance atomically
    if(write_balance_atomic(account_id, initial_balance) == 0) {
        printf("Account %s created with initial balance %d.\n", account_id, initial_balance);
        log_transaction_atomic("Create Account", account_id, "Initial balance", "Success");
    } else {
        printf("Failed to create account %s.\n", account_id);
        log_transaction_atomic("Create Account", account_id, "Initial balance", "Failed");
    }
    pthread_mutex_unlock(&account->lock);
}

// Function to transfer funds atomically
void transfer(const char *from_account_id, const char *to_account_id, int amount) {
    if(strcmp(from_account_id, to_account_id) == 0) {
        printf("Cannot transfer to the same account.\n");
        log_transaction_atomic("Transfer", from_account_id, "Attempted to transfer to self", "Failed");
        return;
    }
    
    // Order accounts to prevent deadlock
    const char *first, *second;
    if(strcmp(from_account_id, to_account_id) < 0) {
        first = from_account_id;
        second = to_account_id;
    } else {
        first = to_account_id;
        second = from_account_id;
    }
    
    Account *first_account = get_account(first);
    Account *second_account = get_account(second);
    
    if(first_account == NULL || second_account == NULL) {
        printf("One or both accounts (%s or %s) do not exist.\n", from_account_id, to_account_id);
        log_transaction_atomic("Transfer", from_account_id, "One or both accounts do not exist", "Failed");
        return;
    }
    
    // Lock both accounts in order
    pthread_mutex_lock(&first_account->lock);
    pthread_mutex_lock(&second_account->lock);
    
    int balance_from, balance_to;
    if(read_balance(from_account_id, &balance_from) != 0 ||
       read_balance(to_account_id, &balance_to) != 0) {
        printf("Error reading account balances.\n");
        log_transaction_atomic("Transfer", from_account_id, "Reading balances failed", "Failed");
        pthread_mutex_unlock(&second_account->lock);
        pthread_mutex_unlock(&first_account->lock);
        return;
    }
    
    if(balance_from < amount) {
        printf("Transfer failed: Insufficient funds in account %s. Current balance: %d\n", from_account_id, balance_from);
        log_transaction_atomic("Transfer", from_account_id, "Insufficient funds", "Failed");
        pthread_mutex_unlock(&second_account->lock);
        pthread_mutex_unlock(&first_account->lock);
        return;
    }
    
    int new_balance_from = balance_from - amount;
    int new_balance_to = balance_to + amount;
    
    // Write new balances atomically
    if(write_balance_atomic(from_account_id, new_balance_from) == 0 &&
       write_balance_atomic(to_account_id, new_balance_to) == 0) {
        printf("Transferred %d from %s to %s.\n", amount, from_account_id, to_account_id);
        log_transaction_atomic("Transfer", from_account_id, "Transfer successful", "Success");
        log_transaction_atomic("Transfer", to_account_id, "Transfer received", "Success");
    } else {
        // Rollback in case of failure
        write_balance_atomic(from_account_id, balance_from);
        write_balance_atomic(to_account_id, balance_to);
        printf("Transfer from %s to %s failed and has been rolled back.\n", from_account_id, to_account_id);
        log_transaction_atomic("Transfer", from_account_id, "Transfer failed and rolled back", "Failed");
        log_transaction_atomic("Transfer", to_account_id, "Transfer failed and rolled back", "Failed");
    }
    
    pthread_mutex_unlock(&second_account->lock);
    pthread_mutex_unlock(&first_account->lock);
}

// Function to deposit funds into an account
void deposit(const char *account_id, int amount) {
    Account *account = get_account(account_id);
    if(account == NULL) {
        printf("Deposit failed: Account %s does not exist.\n", account_id);
        log_transaction_atomic("Deposit", account_id, "Account does not exist", "Failed");
        return;
    }
    
    pthread_mutex_lock(&account->lock);
    int balance;
    if(read_balance(account_id, &balance) != 0) {
        printf("Error reading balance for account %s.\n", account_id);
        log_transaction_atomic("Deposit", account_id, "Reading balance failed", "Failed");
        pthread_mutex_unlock(&account->lock);
        return;
    }
    
    int new_balance = balance + amount;
    if(write_balance_atomic(account_id, new_balance) == 0) {
        printf("Deposited %d to account %s. New balance: %d\n", amount, account_id, new_balance);
        log_transaction_atomic("Deposit", account_id, "Deposit successful", "Success");
    } else {
        printf("Failed to deposit %d to account %s.\n", amount, account_id);
        log_transaction_atomic("Deposit", account_id, "Deposit failed", "Failed");
    }
    pthread_mutex_unlock(&account->lock);
}

// Function to withdraw funds from an account
void withdraw(const char *account_id, int amount) {
    Account *account = get_account(account_id);
    if(account == NULL) {
        printf("Withdrawal failed: Account %s does not exist.\n", account_id);
        log_transaction_atomic("Withdraw", account_id, "Account does not exist", "Failed");
        return;
    }
    
    pthread_mutex_lock(&account->lock);
    int balance;
    if(read_balance(account_id, &balance) != 0) {
        printf("Error reading balance for account %s.\n", account_id);
        log_transaction_atomic("Withdraw", account_id, "Reading balance failed", "Failed");
        pthread_mutex_unlock(&account->lock);
        return;
    }
    
    if(balance < amount) {
        printf("Withdrawal failed: Insufficient funds in account %s. Current balance: %d\n", account_id, balance);
        log_transaction_atomic("Withdraw", account_id, "Insufficient funds", "Failed");
        pthread_mutex_unlock(&account->lock);
        return;
    }
    
    int new_balance = balance - amount;
    if(write_balance_atomic(account_id, new_balance) == 0) {
        printf("Withdrew %d from account %s. New balance: %d\n", amount, account_id, new_balance);
        log_transaction_atomic("Withdraw", account_id, "Withdrawal successful", "Success");
    } else {
        printf("Failed to withdraw %d from account %s.\n", amount, account_id);
        log_transaction_atomic("Withdraw", account_id, "Withdrawal failed", "Failed");
    }
    pthread_mutex_unlock(&account->lock);
}

// Function to view the balance of an account
void view_balance(const char *account_id) {
    Account *account = get_account(account_id);
    if(account == NULL) {
        printf("View balance failed: Account %s does not exist.\n", account_id);
        log_transaction_atomic("View Balance", account_id, "Account does not exist", "Failed");
        return;
    }
    
    pthread_mutex_lock(&account->lock);
    int balance;
    if(read_balance(account_id, &balance) != 0) {
        printf("Error reading balance for account %s.\n", account_id);
        log_transaction_atomic("View Balance", account_id, "Reading balance failed", "Failed");
        pthread_mutex_unlock(&account->lock);
        return;
    }
    
    printf("Account %s Balance: %d\n", account_id, balance);
    log_transaction_atomic("View Balance", account_id, "Balance viewed", "Success");
    pthread_mutex_unlock(&account->lock);
}

// Function to generate a central log of all account balances
void generate_central_log() {
    char central_log_path[100];
    sprintf(central_log_path, "%s/central_log.txt", ACCOUNTS_DIR);
    FILE *log_file = fopen(central_log_path, "w");
    if(log_file == NULL) {
        printf("Error creating central log.\n");
        return;
    }
    fprintf(log_file, "Central Log - Account Balances\n");
    fprintf(log_file, "--------------------------------------------------\n");
    
    pthread_mutex_lock(&global_lock);
    for(int i = 0; i < account_count; i++) {
        pthread_mutex_lock(&accounts[i].lock);
        int balance;
        if(read_balance(accounts[i].account_id, &balance) == 0) {
            fprintf(log_file, "Account: %s, Balance: %d\n", accounts[i].account_id, balance);
        }
        pthread_mutex_unlock(&accounts[i].lock);
    }
    pthread_mutex_unlock(&global_lock);
    
    fclose(log_file);
    printf("Central log created at: %s\n", central_log_path);
}

// Structure representing a user operation
typedef struct {
    char user_id[50];
    char operation[20];
    char target_account[50];
    int amount;
} UserOperation;

// Thread function to perform user operations
void* user_operations(void *arg) {
    UserOperation *op = (UserOperation*)arg;
    if(strcmp(op->operation, "transfer") == 0) {
        transfer(op->user_id, op->target_account, op->amount);
    }
    else if(strcmp(op->operation, "deposit") == 0) {
        deposit(op->user_id, op->amount);
    }
    else if(strcmp(op->operation, "withdraw") == 0) {
        withdraw(op->user_id, op->amount);
    }
    else if(strcmp(op->operation, "view_balance") == 0) {
        view_balance(op->user_id);
    }
    // Simulate delay
    usleep((rand() % 400 + 100) * 1000); // 100 to 500 milliseconds
    free(op);
    pthread_exit(NULL);
}

// Main function
int main() {
    // Ensure the accounts directory exists
    struct stat st = {0};
    if (stat(ACCOUNTS_DIR, &st) == -1) {
        mkdir(ACCOUNTS_DIR, 0700);
    }
    
    // Set the path for the central transaction log
    strcpy(central_transaction_log, ACCOUNTS_DIR "/" TRANSACTION_LOG);
    
    // List of user IDs
    char *user_ids[] = {"User1", "User2", "User3"};
    int num_users = sizeof(user_ids)/sizeof(user_ids[0]);
    
    // Initial balance for each account
    int initial_balance = 1000;
    
    // Create user accounts
    printf("Creating user accounts...\n");
    for(int i = 0; i < num_users; i++) {
        create_account(user_ids[i], initial_balance);
    }
    printf("All accounts created.\n\n");
    
    // Define threads
    pthread_t threads[10];
    int thread_count = 0;
    
    // Define user operations
    // Example: Transfer from User1 to User2, transfer from User2 to User3, transfer from User3 to User1
    UserOperation *op1 = malloc(sizeof(UserOperation));
    strcpy(op1->user_id, "User1");
    strcpy(op1->operation, "transfer");
    strcpy(op1->target_account, "User2");
    op1->amount = 500;
    pthread_create(&threads[thread_count++], NULL, user_operations, (void*)op1);
    
    UserOperation *op2 = malloc(sizeof(UserOperation));
    strcpy(op2->user_id, "User2");
    strcpy(op2->operation, "transfer");
    strcpy(op2->target_account, "User3");
    op2->amount = 300;
    pthread_create(&threads[thread_count++], NULL, user_operations, (void*)op2);
    
    UserOperation *op3 = malloc(sizeof(UserOperation));
    strcpy(op3->user_id, "User3");
    strcpy(op3->operation, "transfer");
    strcpy(op3->target_account, "User1");
    op3->amount = 200;
    pthread_create(&threads[thread_count++], NULL, user_operations, (void*)op3);
    
    // Additional operations can be added here
    // For example:
    // UserOperation *op4 = malloc(sizeof(UserOperation));
    // strcpy(op4->user_id, "User1");
    // strcpy(op4->operation, "deposit");
    // op4->amount = 150;
    // pthread_create(&threads[thread_count++], NULL, user_operations, (void*)op4);
    
    // Wait for all threads to complete
    for(int i = 0; i < thread_count; i++) {
        pthread_join(threads[i], NULL);
    }
    
    // Generate central log after all operations
    generate_central_log();
    printf("All operations completed.\n");
    
    // Destroy mutexes
    for(int i = 0; i < account_count; i++) {
        pthread_mutex_destroy(&accounts[i].lock);
    }
    pthread_mutex_destroy(&global_lock);
    pthread_mutex_destroy(&transaction_log_lock);
    
    return 0;
}
